### PR TITLE
Feature/compiler

### PR DESCRIPTION
**Pull request by virtustron**

### CODE REVIEW

**Review by AndriiSerbeniuk — DISMISSED** [review]
Looks great

**Review by OleksandrStrashnenko — DISMISSED** [review]

**Review by AndriiSerbeniuk — DISMISSED** [review]

File: compiler/src/compilers/cpp_compiler.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        };
+
+        let bin_file_name = generate_filename().to_owned();
+        
+        let mut full_bin_filename = PathBuf::new();
+        full_bin_filename.push(&input_data.compiled_directory_path);
+        full_bin_filename.push(&bin_file_name);
+
+        let compiler_output = Command::new("g++")
+                .args(&[&input_data.compiler_options,
```
Comment (OleksandrStrashnenko) [inline compiler/src/compilers/cpp_compiler.rs]: There is a problem with compiler_options containing "".  
g++ takes "" as a file name.

**Review by AndriiSerbeniuk — DISMISSED** [review]

File: compiler/src/compilers/cpp_compiler.rs (hunk 1 above)
Comment (virtustron) [reply compiler/src/compilers/cpp_compiler.rs]: Thanks a lot! Fix this problem - add if statement

**Review by valentynfaychuk — DISMISSED** [review]

**Review by AndriiSerbeniuk — DISMISSED** [review]

**Review by OleksandrStrashnenko — DISMISSED** [review]

**Review by AndriiSerbeniuk — DISMISSED** [review]

**Review by valentynfaychuk — APPROVED** [review]

**Review by AndriiSerbeniuk — APPROVED** [review]

File: compiler/src/compilers/compiler.rs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+use super::super::data::input_data::InputData;
+use super::super::data::output_data::OutputData;
+
+
+pub trait Compiler {
+    fn compile(&self, input_data: &InputData) -> Result<OutputData, &'static str>;
```
Comment (valentynfaychuk) [inline compiler/src/compilers/compiler.rs:6]: Seems like calls to this function can block for a considerable amount of time, can this function be `async`?

File: compiler/src/storage/name_generator.rs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+/// 
+/// ```
+/// So, ```bin_file_name``` holds ```"file-2021-11-01-08-35-26-518851300"```
+/// 
+
+pub fn generate_filename() -> String {
+    let now = chrono::Utc::now();
+    let filename = String::from(now.format("file-%Y-%m-%d-%H-%M-%S-%f").to_string());
+    filename
+}
```
Comment (valentynfaychuk) [inline compiler/src/storage/name_generator.rs:26]: Can we have "newlines" at the end of the files?

**Review by OleksandrStrashnenko — APPROVED** [review]

**Pull request merged**